### PR TITLE
Use Strides in ViewIndex and MultiIndex

### PR DESCRIPTION
**Pull request by jl-wynen**

### CODE REVIEW

File: variable/include/scipp/variable/variable.h [hunk 1]
```diff
@@ -93,28 +93,28 @@ class SCIPP_VARIABLE_EXPORT Variable {
...
   }
 
-  Variable slice(const Slice params) const;
+  [[nodiscard]] Variable slice(Slice params) const;
   void validateSlice(const Slice &s, const Variable &data) const;
-  [[maybe_unused]] Variable &setSlice(const Slice params, const Variable &data);
+  [[maybe_unused]] Variable &setSlice(Slice params, const Variable &data);
 
-  void rename(const Dim from, const Dim to);
+  void rename(Dim from, Dim to);
```
Comment (SimonHeybrock) [inline variable/include/scipp/variable/variable.h:100]: What is wrong with passing these as const? We don't want to modify them.

File: python/numpy.h [hunk 2]
```diff
@@ -148,14 +144,15 @@ bool memory_overlaps(const py::array_t<T> &data, const View &view) {
   const auto begin = view.begin();
   const auto end = view.end();
   if (begin == end) {
+    // Scalars cannot overlap with numpy arrays in a way that matters to copy.
```
Comment (SimonHeybrock) [inline python/numpy.h:147]: Why does this comment refer to scalars? Scalars should not have `begin==end`?

File: core/test/view_index_test.cpp [hunk 3]
```diff
@@ -27,6 +27,27 @@ class ViewIndex2DTest : public ::testing::Test {
...
     y.add(Dim::Y, ylen);
+
+    s_xy_xy = Strides{xlen, 1};
+    s_xy_yx = Strides{1, ylen};
+    s_x_xy = Strides{1};
+    s_x_yx = Strides{ylen};
+    s_xy_xy_x_edges = Strides{xlen + 1, 1};
+  }
+
+  static Strides make_strides(const Dimensions &iter_dims,
```
Comment (SimonHeybrock) [inline core/test/view_index_test.cpp:38]: unused

File: core/test/view_index_test.cpp [hunk 4]
```diff
@@ -106,30 +132,8 @@ TEST_F(ViewIndex2DTest, increment_2D_transpose) {
   }
 }
 
-TEST_F(ViewIndex2DTest, increment_1D) {
```
Comment (SimonHeybrock) [inline core/test/view_index_test.cpp:109]: So the old behavior was to wrap, and we don't do this any more, so the tests are removed?

File: core/include/scipp/core/element_array_view.h [hunk 5]
```diff
@@ -60,39 +62,40 @@ class ElementArrayViewParams_iterator
 /// Base class for ElementArrayView<T> with functionality independent of T.
 class SCIPP_CORE_EXPORT ElementArrayViewParams {
 public:
-  ElementArrayViewParams(const scipp::index offset, const Dimensions &iterDims,
-                         const Dimensions &dataDims,
-                         const BucketParams &bucketParams);
+  ElementArrayViewParams(scipp::index offset, const Dimensions &iter_dims,
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element_array_view.h:65]: Keep the `const`?

File: core/element_array_view.cpp [hunk 6]
```diff
@@ -48,17 +44,52 @@ ElementArrayViewParams::ElementArrayViewParams(const scipp::index offset,
...
+    for (scipp::index dim = m_iterDims.ndim() - 1; dim >= 0; --dim) {
+      if (m_strides[dim] != expected_stride) {
+        is_contiguous = false;
+        break;
+      }
+      expected_stride *= m_iterDims[m_iterDims.label(dim)];
+    }
+  }
+  if (!is_contiguous)
     throw std::runtime_error("Data is not contiguous");
```
Comment (SimonHeybrock) [inline core/element_array_view.cpp:75]: `DimensionError` may be a better choice?

File: core/element_array_view.cpp [hunk 7]
```diff
@@ -48,17 +44,52 @@ ElementArrayViewParams::ElementArrayViewParams(const scipp::index offset,
...
+    is_contiguous = false;
+  } else {
+    scipp::index expected_stride = 1;
+    for (scipp::index dim = m_iterDims.ndim() - 1; dim >= 0; --dim) {
+      if (m_strides[dim] != expected_stride) {
+        is_contiguous = false;
+        break;
+      }
+      expected_stride *= m_iterDims[m_iterDims.label(dim)];
+    }
```
Comment (SimonHeybrock) [inline core/element_array_view.cpp:72]: Can we just use `m_strides == Strides(m_iterDims)`?

File: common/include/scipp/common/index_composition.h [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+/// @note This function uses *strides* and not a shape, meaning strides[d]
+///       is not the extent of the array in dimension d but rather the step
+///       length in the flat index to advance one element in d.
+///       It is thus *not* complementary to extract_indices.
+template <class ForwardIt1, class ForwardIt2>
+constexpr auto flat_index_from_strides(ForwardIt1 strides_it,
+                                       const ForwardIt2 strides_end,
+                                       ForwardIt2 indices_it) noexcept {
+  std::remove_const_t<std::remove_reference_t<decltype(*strides_it)>> result =
+      0;
```
Comment (SimonHeybrock) [inline common/include/scipp/common/index_composition.h:27]: `std::decay_t` would be shorter?

File: common/include/scipp/common/index_composition.h [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                             ForwardIt2 strides_it) noexcept {
+  scipp::index begin = 0;
+  scipp::index end = 0;
+  for (; shape_it != shape_end; ++shape_it, ++strides_it) {
+    if (*strides_it < 0)
+      begin += *shape_it * *strides_it;
+    else
+      end += *shape_it * *strides_it;
+  }
+  // For scalars the above produces begin==end because all strides == 0.
```
Comment (SimonHeybrock) [inline common/include/scipp/common/index_composition.h:53]: Scalar don't have strides==0, rather no strides? Is `begin == end` the correct check?

File: common/include/scipp/common/index_composition.h [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+/// l_{ndim-1} to allow this case to be distinguishable from an index
+/// to the end.
+///
+/// @param flat_index I
+/// @param ndim n
+/// @param shape {l_d}
+/// @param indices {i_d}
+/// @note This function uses a *shape*, i.e. individual dimension sizes
+///       to encode the size of the array.
+///       It is thus *not* complementary to flat_index_from_strides.
```
Comment (SimonHeybrock) [inline common/include/scipp/common/index_composition.h:87]: You mean "It *is* thus complementary to ...`"?

File: common/test/index_test.cpp [hunk 11]
```diff
@@ -3,6 +3,152 @@
...
+}
+
+TEST(IndexCompositionTest, flat_index_from_strides_3d) {
+  for (scipp::index stride0 = 0; stride0 < 6; ++stride0) {
+    for (scipp::index stride1 = 0; stride1 < 6; ++stride1) {
+      for (scipp::index stride2 = 0; stride2 < 6; ++stride2) {
+        const std::array<scipp::index, 3> strides{stride0, stride1, stride2};
+        for (scipp::index index0 = 0; index0 < stride0; ++index0) {
+          for (scipp::index index1 = 0; index1 < stride1; ++index1) {
+            for (scipp::index index2 = 0; index2 < stride2; ++index2) {
```
Comment (SimonHeybrock) [inline common/test/index_test.cpp:54]: Why does the upper index bound depend on the stride?

File: common/test/index_test.cpp [hunk 12]
```diff
@@ -3,6 +3,152 @@
...
+        }
+      }
+    }
+  }
+}
+
+TEST(IndexCompositionTest, flat_index_from_strides_3d) {
+  for (scipp::index stride0 = 0; stride0 < 6; ++stride0) {
+    for (scipp::index stride1 = 0; stride1 < 6; ++stride1) {
+      for (scipp::index stride2 = 0; stride2 < 6; ++stride2) {
```
Comment (SimonHeybrock) [inline common/test/index_test.cpp:50]: What about negative strides?

File: common/test/index_test.cpp [hunk 13]
```diff
@@ -3,6 +3,152 @@
...
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+TEST(IndexCompositionTest, extract_indices_0d) {
+  const std::array<scipp::index, 3> shape{-1, -2, -3};
```
Comment (SimonHeybrock) [inline common/test/index_test.cpp:68]: What is the meaning of negative values in `shape`?

File: core/include/scipp/core/view_index.h [hunk 14]
```diff
@@ -5,61 +5,62 @@
...
-        m_coord[d] = remainder % m_extent[d];
-        remainder /= m_extent[d];
-      } else {
-        m_coord[d] = 0;
-      }
+  constexpr void set_index(const scipp::index index) noexcept {
+    m_view_index = index;
+    extract_indices(index, m_ndim, m_shape, m_coord);
+    m_memory_index = flat_index_from_strides(
+        m_strides.begin(), m_strides.end(m_ndim), m_coord.begin());
```
Comment (SimonHeybrock) [inline core/include/scipp/core/view_index.h:39]: Have you compared performance to the old approach? It was always very slow and should not be relied on, but wondering if it got any better or worse?

File: core/include/scipp/core/view_index.h [hunk 15]
```diff
@@ -104,15 +105,25 @@ class SCIPP_CORE_EXPORT ViewIndex {
...
-  std::array<scipp::index, NDIM_MAX> m_extent = {};
-  scipp::index m_fullIndex;
-  int32_t m_dims;
-  int32_t m_subdims = 0;
-  std::array<int32_t, NDIM_MAX> m_offsets;
-  std::array<scipp::index, NDIM_MAX> m_factors;
+  /// Shape in iteration dimensions.
+  std::array<scipp::index, NDIM_MAX> m_shape = {};
+  /// Strides in memory.
+  Strides m_strides;
```
Comment (SimonHeybrock) [inline core/include/scipp/core/view_index.h:117]: Do we need `m_strides` AND `m_delta`?

File: core/include/scipp/core/multi_index.h [hunk 16]
```diff
@@ -91,51 +102,57 @@ template <scipp::index N> class MultiIndex {
...
   }
 
-  constexpr void seek_bucket() noexcept {
+  constexpr void seek_bin() noexcept {
     do {
       // go through bin dims which have reached their end (including last
       // pre-bin dim)
       for (scipp::index d = m_ndim_nested - 1;
            (m_coord[d] == m_shape[d]) && (d < NDIM_MAX - 1); ++d) {
+        // Increment early so that we can check whether we need ot load bins.
```
Comment (SimonHeybrock) [inline core/include/scipp/core/multi_index.h:136]: ot -> to

File: core/include/scipp/core/multi_index.h [hunk 17]
```diff
@@ -189,11 +206,15 @@ template <scipp::index N> class MultiIndex {
 
   [[nodiscard]] constexpr scipp::index
   inner_distance_to(const MultiIndex &other) const noexcept {
+    // The end index has coord[0] = 0.
+    if (other.m_coord[0] == 0) {
+      return inner_distance_to_end();
+    }
     return other.m_coord[0] - m_coord[0];
```
Comment (SimonHeybrock) [inline core/include/scipp/core/multi_index.h:213]: Can you explain the reason behind this change?

File: core/test/element_array_view_test.cpp [hunk 18]
```diff
@@ -113,265 +103,236 @@ auto range(const scipp::index end) {
...
 }
 
-TEST(ElementArrayViewTest, bad_broadcast) {
-  Dimensions dims{Dim::X, 2};
-  Dimensions target(Dim::X, 3);
-  EXPECT_THROW(ElementArrayView<int32_t>(range(2).data(), 0, target, dims),
-               except::DimensionError);
-}
-
-TEST(ElementArrayViewTest, broadcast_inner) {
```
Comment (SimonHeybrock) [inline core/test/element_array_view_test.cpp:123]: Why is this test removed? Same question for more tests below.

File: core/test/element_array_view_test.cpp [hunk 19]
```diff
@@ -113,265 +103,236 @@ auto range(const scipp::index end) {
...
+  EXPECT_TRUE(
+      equals(ElementArrayView(range(24).data(), 0, target,
+                              transposed_strides(dims, target.labels())),
+             {0, 4, 8,  12, 16, 20, 1, 5, 9,  13, 17, 21,
+              2, 6, 10, 14, 18, 22, 3, 7, 11, 15, 19, 23}));
 }
 
 TEST(ElementArrayViewTest, collapse_inner) {
-  Dimensions dims{{Dim::X, Dim::Y, Dim::Z}, {2, 3, 4}};
+  // base dims = {{Dim::X, Dim::Y, Dim::Z}, {2, 3, 4}}
```
Comment (SimonHeybrock) [inline core/test/element_array_view_test.cpp:160]: commented code?

File: core/test/element_array_view_test.cpp [hunk 20]
```diff
@@ -113,265 +103,236 @@ auto range(const scipp::index end) {
...
                      {0, 4, 8, 12, 16, 20}));
   // This is a typical use for the offset parameter.
-  EXPECT_TRUE(equals(ElementArrayView(range(24).data(), 3, target, dims),
+  EXPECT_TRUE(equals(ElementArrayView(range(24).data(), 3, target, strides),
                      {3, 7, 11, 15, 19, 23}));
 }
 
 TEST(ElementArrayViewTest, collapse_interior) {
-  Dimensions dims{{Dim::X, Dim::Y, Dim::Z}, {2, 3, 4}};
+  // base dims = {{Dim::X, Dim::Y, Dim::Z}, {2, 3, 4}}
```
Comment (SimonHeybrock) [inline core/test/element_array_view_test.cpp:171]: commented code? More similar cases below.

File: tools/debug/multi_index_helpers.cpp [hunk 21]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Put into scipp::core namespace:
```
Comment (SimonHeybrock) [inline tools/debug/multi_index_helpers.cpp:2]: Maybe you can just provide these snippets in individual header files that can be included directly at the relevant places?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: core/dimensions.cpp [hunk 22]
```diff
@@ -230,17 +230,27 @@ Dimensions intersection(const Dimensions &a, const Dimensions &b) {
...
                  [&dims](auto &dim) { return dims[dim]; });
   return {labels, shape};
 }
+} // namespace
+
+Dimensions transpose(const Dimensions &dims, const std::vector<Dim> &labels) {
+  if (labels.empty()) {
+    std::vector<Dim> default_labels{dims.labels().rbegin(),
+                                    dims.labels().rend()};
+    return transpose_impl(dims, default_labels);
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply core/dimensions.cpp:250]: This behaviour already existed in the previous implementation. The only real change is the const ref in the parameter. The tests include cases where `labels` is empty.

File: variable/test/transform_test.cpp [hunk 23]
```diff
@@ -321,14 +321,25 @@ TEST_F(TransformBinaryTest, var_with_view) {
   EXPECT_EQ(ab, a);
 }
 
-TEST_F(TransformBinaryTest, in_place_self_overlap_without_variance) {
+TEST_F(TransformBinaryTest, in_place_self_overlap_without_variance_1d) {
   auto a = makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{1.1, 2.2});
-  Variable slice_copy(a.slice({Dim::X, 1}));
+  auto slice_copy = copy(a.slice({Dim::X, 1}));
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply variable/test/transform_test.cpp:326]: To avoid dealing with overlaps when computing the reference below.

File: core/dimensions.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/dimensions.cpp:250]: Right, I misread the diff.

File: variable/include/scipp/variable/variable.h (hunk 1 above)
Comment (jl-wynen) [reply variable/include/scipp/variable/variable.h:100]: They are still const in the definition. But it is not necessary to use const in the declaration and clang-tidy warns about it.
This is a question of style, I think. But from what I understand, the general consensus is to avoid const in parameters that are passed by value because it is an implementation detail.

File: variable/test/transform_test.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply variable/test/transform_test.cpp:326]: But overlaps are irrelevant for non-in-place ops?

File: python/numpy.h (hunk 2 above)
Comment (jl-wynen) [reply python/numpy.h:147]: Hm, right. Seems I confused this with how `memory_bounds` used to work before the last fix.

But the implementation should still be correct for scalars. Removing the comment.

File: core/test/view_index_test.cpp (hunk 3 above)
Comment (jl-wynen) [reply core/test/view_index_test.cpp:38]: removed

File: variable/include/scipp/variable/variable.h (hunk 1 above)
Comment (SimonHeybrock) [reply variable/include/scipp/variable/variable.h:100]: I see, people seem to be getting into fights about this. While I agree that it is an implementation detail I feel it may be better to use the "const in both declaration and definition" style? Won't thinks gets messy otherwise if you change things from inline to not inline? It also makes it harder to add new functions unless have you have an IDE that does the "const" removal automatically.

File: python/numpy.h (hunk 2 above)
Comment (SimonHeybrock) [reply python/numpy.h:147]: Also, why is the check required at all? Isn't the condition below sufficient?

File: core/test/view_index_test.cpp (hunk 4 above)
Comment (jl-wynen) [reply core/test/view_index_test.cpp:109]: No. I removed them because I had initially removed the ability to broadcast with ViewIndex. I restored the tests now.

File: variable/test/transform_test.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply variable/test/transform_test.cpp:326]: Right. But it doesn't hurt. Would you rather I removed the copy?

File: variable/include/scipp/variable/variable.h (hunk 1 above)
Comment (jl-wynen) [reply variable/include/scipp/variable/variable.h:100]: > Won't thinks gets messy otherwise if you change things from inline to not inline?

Does not really happen very often.

> It also makes it harder to add new functions unless have you have an IDE that does the "const" removal automatically.

I can't follow. Do you mean copy-pasting the function header between header and implementation files?

I can revert those changes. But then i would disable https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html because inspections are too noisy already.

File: variable/test/transform_test.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply variable/test/transform_test.cpp:326]: I guess, since it confuses me?

File: python/numpy.h (hunk 2 above)
Comment (jl-wynen) [reply python/numpy.h:147]: Yes, I think so

File: core/element_array_view.cpp (hunk 6 above)
Comment (jl-wynen) [reply core/element_array_view.cpp:75]: Not sure. `DimensionError` is usually about extents of data and not its underlying layout.

File: variable/include/scipp/variable/variable.h (hunk 1 above)
Comment (SimonHeybrock) [reply variable/include/scipp/variable/variable.h:100]: Leave it.

File: core/element_array_view.cpp (hunk 7 above)
Comment (jl-wynen) [reply core/element_array_view.cpp:72]: Yes. The current implementation can break out of the loop early but that should not matter for performance.

File: common/include/scipp/common/index_composition.h (hunk 8 above)
Comment (jl-wynen) [reply common/include/scipp/common/index_composition.h:27]: Done

File: common/include/scipp/common/index_composition.h (hunk 9 above)
Comment (jl-wynen) [reply common/include/scipp/common/index_composition.h:53]: True.

It is in principle possible to produce `begin == end` using negative strides. Is this something that can happen in practice?

File: common/include/scipp/common/index_composition.h (hunk 10 above)
Comment (jl-wynen) [reply common/include/scipp/common/index_composition.h:87]: No. The point of the comment is that `extract_indices` uses a shape but `flat_index_from_strides` uses strides. Thus you need to convert between the two if you want to invert one function with the other.

File: common/include/scipp/common/index_composition.h (hunk 9 above)
Comment (SimonHeybrock) [reply common/include/scipp/common/index_composition.h:53]: We should frequently get `begin == end` if a dimension has extent 0. Yes this happens in practice.

If strides are negative I would think we should *not* get `begin == end`, ever, unless extent is also 0?

File: common/include/scipp/common/index_composition.h (hunk 10 above)
Comment (SimonHeybrock) [reply common/include/scipp/common/index_composition.h:87]: Oh, so you "complementary" in the sense of "reverse/inverse", not "set" complementary. Can you use the less ambiguous term?

File: common/test/index_test.cpp (hunk 11 above)
Comment (jl-wynen) [reply common/test/index_test.cpp:54]: Good point. This should use the shape.

File: common/include/scipp/common/index_composition.h (hunk 9 above)
Comment (jl-wynen) [reply common/include/scipp/common/index_composition.h:53]: `begin` and `end` are cummulative over dimensions. If there is one dim with extent 0 but others with greater extend, we get `begin != end`. The only way the two can be equal is if either all dims have extent 0, we have a scalar, or if contriubtions from different dimensions cancel.

Comment (SimonHeybrock) [reply common/include/scipp/common/index_composition.h:53]: Hmm, but for scalars we do *not* want `begin == end`? Shouldn't it be `end = begin + 1` in that case?

Comment (SimonHeybrock) [reply common/include/scipp/common/index_composition.h:53]: I mean, that is what is being done here, but my points is there are legitimate cases with `begin == end`, but scalars are not, so only that case should be fixed here?

File: common/test/index_test.cpp (hunk 13 above)
Comment (jl-wynen) [reply common/test/index_test.cpp:68]: To catch it `extract_indices` overwrites the correct values and nothing else because it should never write negative numbers.

File: common/include/scipp/common/index_composition.h (hunk 9 above)
Comment (SimonHeybrock) [reply common/include/scipp/common/index_composition.h:53]: Maybe just return early on top:
```cpp
if (shape_it == shape_end)  // scalar
  return {0, 1};
```

File: core/include/scipp/core/view_index.h (hunk 14 above)
Comment (jl-wynen) [reply core/include/scipp/core/view_index.h:39]: No. Is there already a benchmark for this?

File: common/test/index_test.cpp (hunk 13 above)
Comment (SimonHeybrock) [reply common/test/index_test.cpp:68]: I don't understand you reply, can you reformulate?

File: core/include/scipp/core/view_index.h (hunk 14 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/view_index.h:39]: Hmm, no, according to a quick look through our benchmarks. I think you simply need to check
```python
auto vals = var.values<double>();
vals[i];
```
for some `i`.

Comment (SimonHeybrock) [reply core/include/scipp/core/view_index.h:39]: There is `element_array_view_benchmark.cpp` which would cover the non-random access, I think?

File: common/test/index_test.cpp (hunk 13 above)
Comment (jl-wynen) [reply common/test/index_test.cpp:68]: I mixed things up. The indices start with negative values to ensure that `extract_indices` overwrites them. Initialising with 0 would not catch the case where 0 is the correct answer but `extract_indices` fails to write it out.
For `shape`, this is just a way or marking unused elements. And maybe trigger an error if the function uses those elements, though that should happen regardless of what value is in there.

File: core/include/scipp/core/view_index.h (hunk 14 above)
Comment (jl-wynen) [reply core/include/scipp/core/view_index.h:39]: So you are explicitly asking about the performance of `set_index`?

File: common/test/index_test.cpp (hunk 13 above)
Comment (SimonHeybrock) [reply common/test/index_test.cpp:68]: 👍

File: core/include/scipp/core/view_index.h (hunk 14 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/view_index.h:39]: Hmm, I suppose both would be good to know, `operator[]()` as well as iteration/incrementing?

File: core/include/scipp/core/view_index.h (hunk 15 above)
Comment (jl-wynen) [reply core/include/scipp/core/view_index.h:117]: Strictly speaking, no. But this should improve performance. Each one is beneficial for only one of sequential and random lookup.
I will benchmark and see if this really matters.

File: core/include/scipp/core/multi_index.h (hunk 17 above)
Comment (jl-wynen) [reply core/include/scipp/core/multi_index.h:213]: Suppose we have a shape of `[4, 3, 2]` and no chunking in transform. A possible begin index index for the inner loop has `coord = [0, 1, 0]` and the corresponding end index has `coord = [0, 2, 0]`. The 'inner distance' between the two is `4` because they cover the whole inner dimension. But the check in line 213 would return 0 which produces an infinite loop.

This is the bug that only appears in a serial build because with chunking, the end index normally does not have `coord[0] = 0`.

Comment (SimonHeybrock) [reply core/include/scipp/core/multi_index.h:213]: But shouldn't we get `indices.in_same_chunk(end, 1) == false` then and not call this function?

File: core/test/element_array_view_test.cpp (hunk 19 above)
Comment (jl-wynen) [reply core/test/element_array_view_test.cpp:160]: This is meant to show what the underlying dimensions would be given the target dims and strides.

File: tools/debug/multi_index_helpers.cpp (hunk 21 above)
Comment (jl-wynen) [reply tools/debug/multi_index_helpers.cpp:2]: The problem is that the `MultiIndex` operator needs to be a friend because there are no accessors for most of the internal state.

I considered putting it into the class and surrounding it with `#ifndef NDEBUG` but I would rather not have conditionally compiled pieces of code.

Comment (SimonHeybrock) [reply tools/debug/multi_index_helpers.cpp:2]: My point is, if you want to use this:

```cpp
class MultiIndex:
  // ...
private:
  #include multi_index_print
};
```

Comment (SimonHeybrock) [reply tools/debug/multi_index_helpers.cpp:2]: Then you don't need to copy the code, you just include it.

Comment (jl-wynen) [reply tools/debug/multi_index_helpers.cpp:2]: Dirty, but possible. This would also need `operator<<` for array. I can make a general debug IO header which provides that and maybe the same for `std::vector`. Would this still fit into `tools`?

Comment (SimonHeybrock) [reply tools/debug/multi_index_helpers.cpp:2]: I guess, yes.

File: core/include/scipp/core/multi_index.h (hunk 17 above)
Comment (jl-wynen) [reply core/include/scipp/core/multi_index.h]: Yes it should and now it does. There was a `+1` missing in `in_same_chunk`. See the latest changes

File: core/test/element_array_view_test.cpp [hunk 24]
```diff
@@ -113,265 +103,265 @@ auto range(const scipp::index end) {
...
-  // Base view with collapsed Y
-  ElementArrayView base(data.data(), 0, target, dims);
-  // Derived view with Y dependence. Since the base view had no Y it is
-  // broadcasted is *not* giving the original data. The application of this is
-  // some operation like var += var.slice(Dim.Y, 0), where we first slice and
-  // then broadcast the result for a subsequent operation.
-  EXPECT_TRUE(equals(ElementArrayView<const int32_t>(base, dims),
-                     {0,  1,  2,  3,  0,  1,  2,  3,  0,  1,  2,  3,
-                      12, 13, 14, 15, 12, 13, 14, 15, 12, 13, 14, 15}));
-}
```
Comment (SimonHeybrock) [inline core/test/element_array_view_test.cpp:367]: 3 more tests that disappeared above?

File: core/test/element_array_view_test.cpp [hunk 25]
```diff
@@ -88,23 +89,12 @@ TEST(ElementArrayView, subview) {
...
-  Dimensions subDims({{Dim::Y, 3}, {Dim::X, 2}});
-  ElementArrayView<double> subView(view, subDims);
-  it = subView.begin();
-  ASSERT_EQ(std::distance(it, subView.end()), 6);
-  EXPECT_EQ(*it++, 0.0);
-  EXPECT_EQ(*it++, 0.0);
-  EXPECT_EQ(*it++, 2.0);
-  EXPECT_EQ(*it++, 2.0);
-  EXPECT_EQ(*it++, 4.0);
-  EXPECT_EQ(*it++, 4.0);
```
Comment (SimonHeybrock) [inline core/test/element_array_view_test.cpp:107]: Why was this removed?

File: core/test/element_array_view_test.cpp (hunk 19 above)
Comment (SimonHeybrock) [reply core/test/element_array_view_test.cpp:189]: But in all cases the comments seems to be a 1:1 copy of the line above? I don't understand which information it provides?

File: core/test/element_array_view_test.cpp (hunk 25 above)
Comment (jl-wynen) [reply core/test/element_array_view_test.cpp]: Don't remember. Restored it

**Comment by jl-wynen** [comment]
Here are benchmark results with a range-for, an iterator based for, and an index based for loop.
- 'old' = main branch
- 'new_base' is what is in this PR
- 'new_nodelta' is what is in the PR except that ViewIndex uses `m_strides` in `increment`

The latter two only really differ in the `rangeFor` benchmark. I am not sure where the difference in the other cases comes from. I would guess that it has to do with the reduced size of `ViewIndex` or reduced setup costs. Though it seems strange that this has an effect for large arrays. Maybe it is just a fluke.

index
iterators
rangefor

**Comment by jl-wynen** [comment]
Transform benchmarks are mostly unaffected. But there seems to be a small improvement for accumulation in an outer dim with small sizes (top pair of lines in top figure)

accum
slice

**Comment by SimonHeybrock** [comment]
Any idea why there is a 20-30% performance drop in the last figure for size>1e9?

**Comment by jl-wynen** [comment]
I find it more strange that the old performance goes up again.

**Review by SimonHeybrock — APPROVED** [review]
